File: src/SnagLite/Services/Resolvers/IResolver.cs
namespace SnagLite.Services.Resolvers;

public sealed record ResolvedMedia(
    string DownloadUrl,
    string Referer,
    string SuggestedName,
    string UserAgent);

public interface IResolver
{
    string Name { get; }
    bool Matches(Uri url);
    Task<ResolvedMedia?> ResolveAsync(Uri url, HttpClient http, CancellationToken ct);
}

File: android/app/src/main/java/com/patron/snaglite/download/resolvers/IResolver.kt
package com.patron.snaglite.download.resolvers

import okhttp3.OkHttpClient

data class ResolvedMedia(
    val downloadUrl: String,
    val referer: String,
    val suggestedName: String,
    val userAgent: String,
)

abstract class IResolver {
    abstract val name: String

    /** Set by `resolve` when it returns null, so callers can surface a useful message. */
    @Volatile
    var lastError: String? = null

    abstract fun matches(host: String): Boolean
    abstract suspend fun resolve(url: String, http: OkHttpClient): ResolvedMedia?
}
